File: src/python/kafka_producer.py

from kafka import SimpleProducer, KafkaClient
from tweepy.streaming import StreamListener
from tweepy import OAuthHandler
from tweepy import Stream
from credentials import consumer_key, consumer_secret, access_token, access_token_secret

topic = b'twitter'
kafka = KafkaClient('localhost:9092')
producer = SimpleProducer(kafka)
keywords = "brexit".split()

class StdOutListener(StreamListener):

    def on_data(self, data):
        print(data)
        time.sleep(1000)
        print("\n\n\n")
        producer.send_messages(topic, data.encode('utf-8'))
        return True

    def on_error(self, status):
        print(status)

if __name__ == '__main__':

    print('producer stream')
    l = StdOutListener()
    auth = OAuthHandler(consumer_key, consumer_secret)
    auth.set_access_token(access_token, access_token_secret)
    stream = Stream(auth, l)

    while True:
        try:
            stream.sample()
            stream.filter(languages=["en"], track=keywords)
        except:
            pass

File: src/ruby/kafka-producer.rb
require 'tweetstream'
require 'ruby-kafka'

def start_tweet_stream tracks
  TweetStream.configure do |config|
    config.consumer_key       =  ENV['consumer_key']
    config.consumer_secret    =  ENV['consumer_secret']
    config.oauth_token        =  ENV['access_token']
    config.oauth_token_secret =  ENV['access_token_secret']
    config.auth_method        =  :oauth
  end

  logger = Logger.new($stderr)
  brokers = ["localhost:9092"]

  topic = "twitter"

  kafka = Kafka.new(brokers, client_id: "twitter-producer", logger: logger)

  producer = kafka.producer

  begin
    TweetStream::Client.new.track(tracks,  language: 'en') do |status|
      p "#{status.user.name}"
      p "#{status.text}"
      print "\n\n\n"
      message = tracks.join + "@B1A2R3R4I5E6R7@" + status.text
      producer.produce(message, topic: topic)
      producer.deliver_messages
  end
    
  ensure
    producer.deliver_messages
    producer.shutdown
  end
end